django/views/decorators/cache.py

```diff
@@ -26,41 +31,73 @@ def cache_page(timeout, *, cache=None, key_prefix=None):
     )
 
 
+@sync_and_async_middleware
 def cache_control(**kwargs):
     def _cache_controller(viewfunc):
-        @wraps(viewfunc)
-        def _cache_controlled(request, *args, **kw):
+        def _process_request(request):
             # Ensure argument looks like a request.
             if not hasattr(request, "META"):
                 raise TypeError(
                     "cache_control didn't receive an HttpRequest. If you are "
                     "decorating a classmethod, be sure to use "
                     "@method_decorator."
                 )
-            response = viewfunc(request, *args, **kw)
+
+        def _process_response(response, **kwargs):
             patch_cache_control(response, **kwargs)
+
+        @wraps(viewfunc)
+        def _wrapper_view_sync(request, *args, **kw):
+            _process_request(request)
+            response = viewfunc(request, *args, **kw)
+            _process_response(response, **kwargs)
+            return response
+
+        @wraps(viewfunc)
+        async def _wrapper_view_async(request, *args, **kw):
+            _process_request(request)
+            response = await viewfunc(request, *args, **kw)
+            _process_response(response, **kwargs)
             return response
 
-        return _cache_controlled
+        if asyncio.iscoroutinefunction(viewfunc):
```

Comment:
Please update to use `asgiref.sync.iscoroutinefunction` (For PY312 compatibility, see #16380)